### PR TITLE
perf: Asset Depreciations and Balances report

### PR DESCRIPTION
**Pull request by Don-Leopardo**
### The issue:
This report took ~5 min to be generated and the server was unresponsive in the meantime.

### The filters:
Every category in the last year.

### The data:
~1400 assets distributed among 76 assets categories with a total of ~243000 depreciation schedules.

### The solution:
I've converted the python logic to sql. Now, MariaDB returns all data almost ready to print.

### The results:
Now, the report takes ~22 sec the first time it's executed and then, if you change the filters it takes ~1 sec. The server does not die.

**Note:** The report had a nested loop, it was doing some calculations for each schedule of each asset. Moreover, it was using `frappe.get_doc` with each asset, instead of `frappe.get_value` (with just the necessary values). Anyway, all of that is a thing from the past, now almost everything is done in sql.

### CODE REVIEW

**Comment by Don-Leopardo** [comment]
@SaiFi0102 first of all, sorry for tagging you, but this was labeled as Accounts and you are the most active accountant group member on github (or at least the one that i always see).
I wanted to ask you if you can review this pr, or if you can assign someone to it. Thank you.

**Comment by SaiFi0102** [comment]
@Don-Leopardo Sure Don, I can review the code, but I don't have the data to test the correctness of the data.

File: erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py
Comment (Don-Leopardo) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: It can be, look at this:
Captura de pantalla 2019-05-13 a la(s) 08 30 53
Captura de pantalla 2019-05-13 a la(s) 08 30 44
Captura de pantalla 2019-05-13 a la(s) 08 30 30

Comment (Don-Leopardo) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: @barredterra Is it possible to dismiss a change request? Or what should i do in this case?

Comment (barredterra) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: I see, very weird. 

You should be able to mark the request as resolved. But no worries, my review doesn't count as i'm not on the frappe team.

Comment (Don-Leopardo) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: @barredterra I was wanting to know the "general behavior" for this kind of situations. Should I "resolve conversation" before you answer my answer of should I wait for your confirmation?

Comment (barredterra) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: I think you can resolve right away when you're sure that the problem is resolved.

File: erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py
Comment (rmehta) [inline erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: fix indentation. Its annoying to scroll in 2 directions to read code!

File: erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py
Comment (rmehta) [inline erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: fix indentation

https://github.com/frappe/erpnext/wiki/Coding-Standards#use-simple-indenting

**Review by rmehta — CHANGES_REQUESTED** [review]
👍 

please fix indentation and send parallel PR for `develop` branch

File: erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py
Comment (Don-Leopardo) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: That was a horrible indentation, sorry.

File: erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py
Comment (Don-Leopardo) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: I think that it's fixed. And I made the PR for develop. #18022

File: erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py
Comment (rohitwaghchaure) [reply erpnext/accounts/report/asset_depreciations_and_balances/asset_depreciations_and_balances.py]: @Don-Leopardo the date can't be zero better you can check the condition as below
ifnull(disposal_date, '2133-03-02') >= %(from_date)s

**Comment by rohitwaghchaure** [comment]
@Don-Leopardo you have wrote the sql query which is difficult to read for the new person who will be working on it in a future. Can you break down that query into multiple parts?

**Comment by rohitwaghchaure** [comment]
Also in v11 prepared report feature was introduced which will generate the report at the backend and once the report is generated then it will display in the screen

**Comment by Don-Leopardo** [comment]
@rohitwaghchaure The SQL query cannot be broke down. Except in a graphical form, like, splitting it in different strings to comment in between, if that's what you mean it can be done, but I think that it is worst to understand it that way.

For the other comment, yes, there is the prepared report feature, I know it, but, with it, this report is still going to consume a lot of resources and is still going to demand a lot of time, but in the background. That's not a solution, it's like hiding the problem. And I'm not sure if it hides it completely, because it's probably that the server still freeze.

**Pull request closed without merging**
